src/MatrixFields/matrix_field_utils.jl

```diff
@@ -0,0 +1,141 @@
+    matrix_ld <= 0 && matrix_ud >= 0 ||
+        error("BandedMatrices.jl does not yet support matrices that have \
+               diagonals with indices in the range $matrix_ld:$matrix_ud")
+
+    return n_rows, n_cols, matrix_ld, matrix_ud
+end
+
+"""
+    column_field2array(field)
+
+Converts a field defined on a `FiniteDifferenceSpace` into a `Vector` or a
+`BandedMatrix`, depending on whether or not the elements of the field are
+`BandMatrixRow`s. This involves copying the data stored in the field. Because
+`BandedMatrix` does not currently support operations with `CuArray`s, all GPU
+data is copied to the CPU.
+"""
+function column_field2array(field::Fields.FiniteDifferenceField)
+    if eltype(field) <: BandMatrixRow # field represents a matrix
+        n_rows, n_cols, matrix_ld, matrix_ud = banded_matrix_info(field)
+        matrix = BandedMatrix{eltype(eltype(field))}(
+            undef,
+            (n_rows, n_cols),
+            (-matrix_ld, matrix_ud),
+        )
+        for (index_of_field_entry, matrix_d) in enumerate(matrix_ld:matrix_ud)
+            matrix_diagonal = view(matrix, band(matrix_d))
+            diagonal_field = field.entries.:($index_of_field_entry)
+            diagonal_data =
+                vec(reinterpret(eltype(eltype(field)), parent(diagonal_field)'))
+
+            # Find the rows for which diagonal_data[row] is in the matrix.
+            # Note: The matrix index (1, 1) corresponds to the diagonal index 0,
+            # and the matrix index (n_rows, n_cols) corresponds to the diagonal
+            # index n_cols - n_rows.
+            first_row = matrix_d < 0 ? 1 - matrix_d : 1
+            last_row = matrix_d < n_cols - n_rows ? n_rows : n_cols - matrix_d
+
+            diagonal_data_view = view(diagonal_data, first_row:last_row)
+            @allowscalar copyto!(matrix_diagonal, diagonal_data_view)
+        end
+        return matrix
+    else # field represents a vector
+        return @allowscalar Array(column_field2array_view(field))
+    end
+end
+
+"""
+    column_field2array_view(field)
+
+Similar to `column_field2array(field)`, except that this version avoids copying
+the data stored in the field.
+"""
+function column_field2array_view(field::Fields.FiniteDifferenceField)
+    if eltype(field) <: BandMatrixRow # field represents a matrix
+        _, n_cols, matrix_ld, matrix_ud = banded_matrix_info(field)
+        field_data_transpose =
+            reinterpret(eltype(eltype(field)), parent(field)')
+        matrix_transpose =
+            _BandedMatrix(field_data_transpose, n_cols, matrix_ud, -matrix_ld)
+        return permutedims(matrix_transpose)
+        # TODO: Despite not copying any data, this function still allocates a
+        # small amount of memory because of _BandedMatrix and permutedims.
+    else # field represents a vector
+        return vec(reinterpret(eltype(field), parent(field)'))
+    end
+end
+
+all_columns(::Fields.FiniteDifferenceField) = (((1, 1), 1),)
```

Comment:
Is this supposed to be `::Fields.ColumnField`?

Reply:
It turns out that specializing on `Fields.ColumnField` and `Fields.ExtrudedFiniteDifferenceField` results in a method ambiguity, since the first is specializing on the data container type and the second is specializing on the space type. In order to avoid failing the Aqua tests, the specialization needs to be on `Fields.FiniteDifferenceField` and `Fields.ExtrudedFiniteDifferenceField`.

Reply:
So, `all_columns(::Fields.FiniteDifferenceField) = (((1, 1), 1),)` was written generically to avoid a method ambiguity?

Reply:
Yes. And I also wanted to make sure that the only two valid options are `Fields.ColumnField` and `Fields.ExtrudedFiniteDifferenceField`, since a `BandMatrixRow` field on any other space cannot be interpreted as a matrix.

Reply:
Method ambiguities can be fixed by defining the intersecting case, so, I think what we want is: 
```julia
all_columns(::DataLayouts.DataColumn, ::Spaces.AbstractSpace) = (((1, 1), 1),)
all_columns(::DataLayouts.AbstractData, space::Spaces.ExtrudedFiniteDifferenceSpace) =
    Spaces.all_nodes(Spaces.horizontal_space(space))
all_columns(::DataLayouts.DataColumn, ::Spaces.ExtrudedFiniteDifferenceSpace) = (((1, 1), 1),)
```
This last one should fix the ambiguity.

Reply:
This way the code paths are a bit more explicit / clear.

Reply:
(at the cost of verbosity)

Reply:
Isn't a field on a FiniteDifferenceSpace always going to be stored on a DataColumn, though? That's the assumption I made when writing [this code](https://github.com/CliMA/ClimaCore.jl/blob/eef71dc05a0672d1736e128aeb500c4a18e57727/src/Operators/thomas_algorithm.jl#L23-L49) for the Thomas algorithm. I don't see why we need to handle a column of data stored on a generic AbstractSpace.

Reply:
I think it depends how it's sliced. We can use `bycolumn` to make column views, or `byslab` to make slab views-- those will have different data layouts

Reply:
I'm not making the argument that this will ever be done, I think it's just more defensive/explicit

Reply:
Don't `bycolumn` and `byslab` also change the space, though? If someone were to call `column_field2array` from a `byslab`, the correct behavior would be to throw an error, since it is not possible to interpret a horizontal slab of `BandMatrixRows` as a matrix.

Reply:
With the version you suggested, the `all_columns` function wouldn't throw an error in that case, but the `column_field2array` function still would.

Reply:
If I make the changes you mentioned, how should I change the following condition used lower down in this file:
```
    is_matrix_field =
        field isa Fields.FiniteDifferenceField ||
        field isa Fields.ExtrudedFiniteDifferenceField
```
If a field of `BandMatrixRow`s isn't located on one of these two spaces, then it can't be interpreted as a matrix, and it needs to be printed differently. Should I use `ColumnField` instead of `FiniteDifferenceField` here?

Reply:
> Isn't a field on a FiniteDifferenceSpace always going to be stored on a DataColumn, though?

Not based on its definition: 
```julia
const FiniteDifferenceField{V, S} =
    Field{V, S} where {V <: AbstractData, S <: Spaces.FiniteDifferenceSpace}
```

I'll take a look at `column_field2array` and the `is_matrix_field` part today.